contracts/YieldEulerLever.sol

```diff
@@ -0,0 +1,390 @@
+    ///     amount when repaying.
+    /// @param data The data we encoded for the functions. Here, we only check
+    ///     the first byte for the router.
+    function onFlashLoan(
+        address initiator,
+        address, // The token, not checked as we check the lender address.
+        uint256 borrowAmount,
+        uint256 fee,
+        bytes calldata data
+    ) external override returns (bytes32) {
+        Operation status = Operation(uint256(uint8(data[0])));
+        bytes6 seriesId = bytes6(data[1:7]);
+        bytes6 ilkId = bytes6(data[7:13]);
+        bytes12 vaultId = bytes12(data[13:25]);
+        // We trust the lender, so now we can check that we were the initiator.
+        if (initiator != address(this)) revert FlashLoanFailure();
+
+        // Decode the operation to execute and then call that function.
+        if (status == Operation.BORROW) {
+            uint128 baseAmount = uint128(uint128(bytes16(data[25:41])));
+            uint256 minCollateral = uint128(bytes16(data[41:57]));
+            borrow(
+                ilkId,
+                seriesId,
+                vaultId,
+                borrowAmount,
+                fee,
+                baseAmount,
+                minCollateral
+            );
+        } else if (status == Operation.REPAY) {
+            uint128 ink = uint128(bytes16(data[25:41]));
+            uint128 art = uint128(bytes16(data[41:57]));
+            repay(
+                ilkId,
+                seriesId,
+                vaultId,
+                uint128(borrowAmount + fee),
+                ink,
+                art
+            );
+        } else if (status == Operation.CLOSE) {
+            uint128 ink = uint128(bytes16(data[25:41]));
+            uint128 art = uint128(bytes16(data[41:57]));
+
+            close(ilkId, vaultId, ink, art);
+        }
+
+        return FLASH_LOAN_RETURN;
+    }
+
+    /// @notice This function is called from within the flash loan. The high
+    ///     level functionality is as follows:
+    ///         - We have supplied and borrowed FYWeth.
+    ///         - We convert it to StEth and put it in the vault.
+    ///         - Against it, we borrow enough FYWeth to repay the flash loan.
+    /// @param ilkId The id of the ilk being borrowed.
+    /// @param seriesId The pool (and thereby series) to borrow from.
+    /// @param vaultId The vault id to put collateral into and borrow from.
+    /// @param borrowAmount The amount of FYWeth borrowed in the flash loan.
+    /// @param fee The fee that will be issued by the flash loan.
+    /// @param baseAmount The amount of own collateral to supply.
+    /// @param minCollateral The final amount of collateral to end up with, or
+    ///     the function will revert. Used to prevent slippage.
+    function borrow(
+        bytes6 ilkId,
+        bytes6 seriesId,
+        bytes12 vaultId,
+        uint256 borrowAmount,
+        uint256 fee,
+        uint128 baseAmount,
```

Comment:
```suggestion
        uint256 baseAmount,
```
Internal functions should accept only uint256 as parameters, and cast to whatever is needed when calling external contracts or storing structs.